### PR TITLE
Static region for secret manager and parametar store

### LINKED ISSUE #578: Add option to set static region for Parameter Store and Secrets Manager integration
Since Parameter Store and Secrets Manager integrations are not able to respect region configuration provided in `application.properties` we should allow users to set region explicitly for both Secrets Manager and Parameter store in `bootstrap.properties` file. This can be done by simply adding `region` property to `@ConfigurationProperties` classes and respect this property when creating clients.

It will fix #477

### PR DESCRIPTION
**Pull request by MatejNedic**
Closes #578

### CODE REVIEW

**Comment by eddumelendez** [comment]
@MatejNedic thanks for your contribution. LGTM

Can you add some tests?

**Comment by MatejNedic** [comment]
Hi @eddumelendez ,
ofc I can. Can you check them out since I am not sure if this is best way to implement them. Do you have any recommendations mb?

File: spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+				.getDeclaredMethod("ssmClient", AwsParamStoreProperties.class);
+		SSMClientMethod.setAccessible(true);
+		AWSSimpleSystemsManagementClient awsSimpleClient = (AWSSimpleSystemsManagementClient) SSMClientMethod
+				.invoke(bootstrapConfig, awsParamStoreProperties);
+
+		Method signingRegionMethod = AmazonWebServiceClient.class.getDeclaredMethod("getSigningRegion");
+		signingRegionMethod.setAccessible(true);
+		String signedRegion = (String) signingRegionMethod.invoke(awsSimpleClient);
+
+		Assert.assertEquals(signedRegion, region);
```
Comment (eddumelendez) [inline spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java:54]: use static import for `assertEquals`

File: spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * Unit test for {@link AwsParamStoreBootstrapConfiguration}.
+ * @author Matej Nedic
+ */
+public class AwsParamStoreBootstrapConfigurationTest {
+
+	AwsParamStoreBootstrapConfiguration bootstrapConfig = new AwsParamStoreBootstrapConfiguration();
+
+	@Test
+	public void setRegion() throws NoSuchMethodException, InvocationTargetException,
```
Comment (eddumelendez) [inline spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java:38]: I will rename it to `testWithStaticRegion`

File: spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+				.invoke(bootstrapConfig, awsParamStoreProperties);
+
+		Method signingRegionMethod = AmazonWebServiceClient.class.getDeclaredMethod("getSigningRegion");
+		signingRegionMethod.setAccessible(true);
+		String signedRegion = (String) signingRegionMethod.invoke(awsSimpleClient);
+
+		Assert.assertEquals(signedRegion, region);
+	}
+
+	private static final class AwsParamStorePropertiesBuilder {
```
Comment (eddumelendez) [inline spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java:57]: no need to do the builder IMO

File: spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	AwsParamStoreBootstrapConfiguration bootstrapConfig = new AwsParamStoreBootstrapConfiguration();
+
+	@Test
+	public void setRegion() throws NoSuchMethodException, InvocationTargetException,
+			IllegalAccessException {
+		String region = "us-east-2";
+		AwsParamStoreProperties awsParamStoreProperties = new AwsParamStorePropertiesBuilder()
+				.withRegion(region).build();
+
+		Method SSMClientMethod = AwsParamStoreBootstrapConfiguration.class
```
Comment (eddumelendez) [inline spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java:44]: May use `ReflectionUtils` from spring framework to do the reflection job

File: spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java (hunk 1 above)
Comment (MatejNedic) [reply spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java]: Done

File: spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java (hunk 2 above)
Comment (MatejNedic) [reply spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java]: Done

File: spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java (hunk 3 above)
Comment (MatejNedic) [reply spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java]: Done. Agree there shouldn't be builder for only 1 variable.

File: spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java (hunk 4 above)
Comment (MatejNedic) [reply spring-cloud-starter-aws-parameter-store-config/src/test/java/org/springframework/cloud/aws/autoconfigure/paramstore/AwsParamStoreBootstrapConfigurationTest.java]: Done.

**Comment by eddumelendez** [comment]
@MatejNedic thank you so much for your contribution one more time and for taking care of the comments. Your PR has been merged.

**Comment by willome** [comment]
@MatejNedic @eddumelendez : I think it would be better to mimic what has been done with `spring-cloud-starter-aws-secrets-manager-config` : having `spring-cloud-aws-autoconfigure` has a transittive dependency could be probably enough. What do you think ?

**Comment by MatejNedic** [comment]
Hi @willome,
since both param and secret stores are configured during the bootstrap phase rather than regular application context they don't use ContextRegionProviderAutoConfiguration.
For more info, you can check the discussion in #477 .

**Pull request closed without merging**